### PR TITLE
[FIX] Fix flaky test when testing decomposition estimators

### LINKED ISSUE #6542: :rotating_light: NIGHTLY DEPENDENCIES TEST: failure
The run `33517904800` of the workflow NIGHTLY DEPENDENCIES TEST failed.

You can view the report here: https://github.com/nilearn/nilearn/actions/runs/33517904800

Pinging @nilearn/core

### PR DESCRIPTION
**Pull request by hndgzkn**
- Closes #6542 

Use of AI tools:
- [ ] LLM tools were used to generate at least part of the content of this pull-request.
- [x] No LLM tools were used to generate any part of the content of this pull-request.

Changes proposed in this pull request:

- Instantiate estimators in `nilearn.decomposition.tests.test_decomposition_estimators` with random_state to have more deterministic results
- Remove `flaky` mark from `nilearn.decomposition.tests.test_decomposition_estimators.test_check_estimator_nilearn`

Nilearn estimator checks clone the same estimator many times to run the tests. When no `random_state` is specified for the estimator, the clone might have different random_state depending on the order of the tests run. Setting random_state, it is expected to have more deterministic results.

### CODE REVIEW

File: nilearn/maskers/nifti_maps_masker.py [hunk 1]
```diff
@@ -679,11 +678,11 @@ def transform_single_imgs(self, imgs, confounds=None, sample_mask=None):
 
         # we update some attributes
         # that may have been changed by resampling or masking
-        if len(extracted_maps) != self.n_elements_:
-            self.n_elements_ = len(extracted_maps)
-            maps_data = get_data(maps_img_)[:, :, :, extracted_maps]
-            self.maps_img_ = new_img_like(self.maps_img_, maps_data)
-            self._reporting_data["maps_image"] = self.maps_img_
```
Comment (hndgzkn) [inline nilearn/maskers/nifti_maps_masker.py:686]: @Remi-Gau I am working on flaky estimator checks in decomposition module. After setting `random_state`, the test still complains about the shape of input to `inverse_transform`.

In the below piece of code, we create the list of methods of the estimator to test:

https://github.com/nilearn/nilearn/blob/c537ff84da68eea1929b9bb7b089f97cca39e682/nilearn/_utils/estimator_checks.py#L1651-L1666 .

For decomposition, the list is `[transform, predict, score, inverse_transform]`. The input signal for `inverse_transform` is generated using `fitted_estimator.n_elements_` at line 1663. 

As the selected code modifies `fitted_estimator.n_elements_` value after calling transform and then `inverse_transform` is called with the signal that is produced with the value of `n_elements_` before transform, the tests fail. 

At this point I realized that this piece of code is part of transform. So normally it should not change the estimator, non? The comments below actually tells this:

https://github.com/nilearn/nilearn/blob/c537ff84da68eea1929b9bb7b089f97cca39e682/nilearn/maskers/nifti_maps_masker.py#L525-L540.

- This part seems to be introduced with #6551 and we both missed that point in [our discussion](https://github.com/nilearn/nilearn/pull/6551#discussion_r3968052217). 

At this point I have a few questions:

- Do you agree with the above?
- If you agree so, is it ok that I revert those changes in this PR?
- For the reporting part, actually what should reports contain, fit time or transform time data?

Comment (bthirion) [reply nilearn/maskers/nifti_maps_masker.py:686]: Sounds right, thx !

Comment (Remi-Gau) [reply nilearn/maskers/nifti_maps_masker.py:686]: actually not quite

the change in that PR was intentional but I totally forgot to update the comment

## why this was needed eventhough this breaks the sklearn convention of 'transform should not change estimator'

The thing to keep in mind is that we want to make sure that when we apply the transform of an estimator to an image and get a numpy array, we must be able to then pass this array to the inverse_transform of the same instance of that estimator and get an image back (without errors).

`inverse_transform` does a check to make sure that the input array has the correct shape, for maps masker that means that the number of column in the array must match the number of "valid" maps that were used for the data extraction at transform time (the attribute `n_elements_` in nilearn - `n_features_in_` in sklearn).

But at transform time masking or resampling can reduce the number of "valid" maps in the masker, so **we need to keep track of this** at least via the `n_elements_`:
 - actually in #6577 I am working on making the behavior of the maps masker more like the one of the label masker where the modified maps image is stored in the different attribute created at fit time.

Note that the NiftiLabelsMasker already behaves like this.

Comment (Remi-Gau) [reply nilearn/maskers/nifti_maps_masker.py:686]: OK I checked and some of the tests in place that are failing are actually testing this behavior and failing: https://github.com/nilearn/nilearn/actions/runs/34476721466/job/102869257386?pr=6578#step:8:17920

Comment (hndgzkn) [reply nilearn/maskers/nifti_maps_masker.py:686]: Yes, but they fail just because I commented out this part of the code. 

If that was not the case, it is estimator checks which fail. Because the function in the above code from estimator checks, fits the estimator, then creates two lists; one containing method name to test, the other input data to method to test. 

Input signal to `inverse_transform` is generated according to `fitted_estimator.n_elements_`. But `inverse_transform` is tested after `transform` is run on the estimator, thus `n_elements_` value has changed. So the input signal for `inverse_transform` becomes an invalid input, and tests fail. 

It is possible to fix it without commenting out these lines in transform.

However I have doubts about changing the state of estimator during transform. For example if I transform `image1` and then `image2`, I can only `inverse_transform` `image1` after I transform it. Otherwise it will not be possible after transforming `image2` if a resampling is done during transforming `image2`.

Comment (hndgzkn) [reply nilearn/maskers/nifti_maps_masker.py:686]: I've checked scikit learn, that might not be the best example, but [here](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) for example it is documented that if `clip` is specified as `True`, inverse transform might not be able to restore original data.

We could may be adopt a similar approach and document possible conditions that `inverse_transform` might not work.

Comment (Remi-Gau) [reply nilearn/maskers/nifti_maps_masker.py:686]: > However I have doubts about changing the state of estimator during transform. For example if I transform `image1` and then `image2`, I can only `inverse_transform` `image1` after I transform it. Otherwise it will not be possible after transforming `image2` if a resampling is done during transforming `image2`.

in most cases in practice I think people use images that have same field of view (shape and affine), so that problem would be rare though it exists: in this case the user has to refit a new instance

note that if we decide for this behavior, then let's be internally consistent and also make sure that nifti label masker behaves the same way: will flag this as something to discuss with the rest of the team

also if we decide for this behavior we should update the warnings that are thrown when maps or labels are dropped at transform time: the users should know they are getting some arrays that are not "inverse transformable"

Comment (Remi-Gau) [reply nilearn/maskers/nifti_maps_masker.py:686]: > It is possible to fix it without commenting out these lines in transform.

how exactly? I think I am missing something

Comment (Remi-Gau) [reply nilearn/maskers/nifti_maps_masker.py:686]: note: all this is orthogonal to the flaky test, no?

Comment (hndgzkn) [reply nilearn/maskers/nifti_maps_masker.py:686]: > note: all this is orthogonal to the flaky test, no?

I think this is the reason why decomposition flaky test fails.

Comment (hndgzkn) [reply nilearn/maskers/nifti_maps_masker.py:686]: > > It is possible to fix it without commenting out these lines in transform.
> 
> how exactly? I think I am missing something

I meant, we can keep changing the estimator state in transform time, and I can fix the error in estimator checks by providing appropriate input to inverse_transform. Let me try.

Comment (hndgzkn) [reply nilearn/maskers/nifti_maps_masker.py:686]: > in most cases in practice I think people use images that have same field of view (shape and affine), so that problem would be rare though it exists: in this case the user has to refit a new instance
> 
> note that if we decide for this behavior, then let's be internally consistent and also make sure that nifti label masker behaves the same way: will flag this as something to discuss with the rest of the team
> 
> also if we decide for this behavior we should update the warnings that are thrown when maps or labels are dropped at transform time: the users should know they are getting some arrays that are not "inverse transformable"

I agree with all :)

Comment (Remi-Gau) [reply nilearn/maskers/nifti_maps_masker.py:686]: > I think this is the reason why decomposition flaky test fails.

I doubt that: the tests have been flaky way before this change was introduced

Comment (hndgzkn) [reply nilearn/maskers/nifti_maps_masker.py:686]: There is also the random_state which has not been set for decomposition estimators. I think that would solve; but might be wrong, couldn't test for all in local. 

I made a trial fix. Code can be improved, just to see if it fixes.

File: nilearn/_utils/estimator_checks.py [hunk 2]
```diff
@@ -1630,66 +1630,70 @@ def check_img_estimator_pickle(estimator_orig) -> None:
...
+        # clone and fit estimator
+        estimator = clone(estimator_orig)
+        X, y = generate_data_to_fit(estimator)
+        set_random_state(estimator)
+        if isinstance(estimator, NiftiSpheresMasker):
+            # NiftiSpheresMasker needs mask_img to run inverse_transform
+            mask_img = new_img_like(X, np.ones(X.shape[:3]))
+            estimator.mask_img = mask_img
+        fitted_estimator = fit_estimator(estimator)
+        return fitted_estimator, X, y
```
Comment (hndgzkn) [inline nilearn/_utils/estimator_checks.py:1643]: This is to have a fresh fitted estimator to test each method.

File: nilearn/_utils/estimator_checks.py [hunk 3]
```diff
@@ -1630,66 +1630,70 @@ def check_img_estimator_pickle(estimator_orig) -> None:
...
+        "inverse_transform",
+    ]:
+        if hasattr(estimator_orig, method):
+            fitted_estimator, X, y = clone_and_fit_estimator(estimator_orig)
+
+            if method == "transform":
+                input_data = (
+                    [X]
+                    if isinstance(
+                        estimator_orig, (SearchLight, _BaseDecomposition)
```
Comment (hndgzkn) [inline nilearn/_utils/estimator_checks.py:1659]: Normally _BaseDecomposition was not there but in local I got some errors that did not exist with the original code.

**Comment by hndgzkn** [comment]
@Remi-Gau I think I might be wrong about my solution. Some tests still fail, but I could not understand if the fix that I introduced worked or not. 

Working with `estimator_checks`, I think this part needs simplification. It is not possible to reproduce the failures locally and navigating in the module or debugging requires too much effort and running small pieces of tests is not straight forward. I'd suggest to simplify it. How does it sound to you?

If you are ok with that, I would suggest to put `estimator_checks` in a separate folder, may be two different modules; `estimator_checks_nilearn.py` and `estimator_checks_sklearn.py`.

Then I would suggest that each function, for example, `check_doc_attributes` can be parametrized with the estimators that has to run this test. This way we can clearly see which estimators have which checks rather than navigating through if conditions to identify those.

If you are ok with that, I would volunteer to draft a PR on that.

**Comment by Remi-Gau** [comment]
> I would suggest to put `estimator_checks` in a separate folder, may be two different modules; `estimator_checks_nilearn.py` and `estimator_checks_sklearn.py`

yes that should already simplify things a bit

**Comment by Remi-Gau** [comment]
> `check_doc_attributes` can be parametrized with the estimators that has to run this test. This way we can clearly see which estimators have which checks rather than navigating through if conditions to identify those.

not sure I like that approach, because then the reason why a certain check has to be run becomes implicit:
- say we add a new type of masker that only works for surface images: now we have to figure out all the checks where to add this new estimator
- or more likely we add a new check: now we have to list all the estimators that this new check applies for

I agree that the current system is hard to navigate but it relies (mostly) on the 'tags' of an estimator to determine what behavior it should comply to: unless I have misunderstood, I get the feeling that your solution would make that link less explicit. But maybe I did not get what you meant. Let's talk later. 😉

**Comment by hndgzkn** [comment]
Following our discussion, I'll open a PR to first move `estimator_checks` to a separate directory from `nilearn._utils`and separate `scikit-learn` and `nilearn` checks.

I am looking at how scikit-learn does these checks as you pointed me. I've came across this [template project](https://github.com/scikit-learn-contrib/project-template/blob/main/skltemplate/tests/test_common.py). I thought may be we can run all estimator checks in `nilearn.tests.test_common` file as shown.

If we were to use `all_estimators` function, some estimators require some construction parameters to be able to run fit, for example [NiftiLabelsMasker](https://github.com/nilearn/nilearn/blob/bc19044243f0071d9cd695ae0f61c579c6ec8586/nilearn/maskers/nifti_labels_masker.py#L476) requires `labels_img` to be specified to be able to run `fit`. 

- I was wondering if there would be an inconvenience to change the code to set `labels_img_` by resampling if it is `labels_img=None` at init time?

- How does it sound to you to check all other estimators and make `fit` independent of construction parameters for them?

**Comment by bthirion** [comment]
> 
>     * I was wondering if there would be an inconvenience to change the code to set `labels_img_` by resampling if it is `labels_img=None` at init time?
> 
I'm not familiar with that code: what happens when  `labels_img=None` at init ?  Doesn't this lead to poor behavior in all circumstances ?

**Comment by Remi-Gau** [comment]
> > ```
> > * I was wondering if there would be an inconvenience to change the code to set `labels_img_` by resampling if it is `labels_img=None` at init time?
> > ```
> 
> I'm not familiar with that code: what happens when `labels_img=None` at init ? Doesn't this lead to poor behavior in all circumstances ?

I think that labels_img=None at init should give an error at fit time for nifti label masker, no?
If it does not, then it should.

**Comment by bthirion** [comment]
> > > ```
> > > * I was wondering if there would be an inconvenience to change the code to set `labels_img_` by resampling if it is `labels_img=None` at init time?
> > > ```
> > 
> > 
> > I'm not familiar with that code: what happens when `labels_img=None` at init ? Doesn't this lead to poor behavior in all circumstances ?
> 
> I think that labels_img=None at init should give an error at fit time for nifti label masker, no? If it does not, then it should.

But then why not get rid of `labels_img=None` ?

**Comment by hndgzkn** [comment]
> I think that labels_img=None at init should give an error at fit time for nifti label masker, no? If it does not, then it should.

You are right, it gives an error. I was wondering if we could change this behavior, and be able to run `fit` when initialized with default parameter values.

**Comment by Remi-Gau** [comment]
> > > > ```
> > > > * I was wondering if there would be an inconvenience to change the code to set `labels_img_` by resampling if it is `labels_img=None` at init time?
> > > > ```
> > > 
> > > 
> > > I'm not familiar with that code: what happens when `labels_img=None` at init ? Doesn't this lead to poor behavior in all circumstances ?
> > 
> > 
> > I think that labels_img=None at init should give an error at fit time for nifti label masker, no? If it does not, then it should.
> 
> But then why not get rid of `labels_img=None` ?

sklearn convention: estimator must be instantiable by default, so all parameters must have a default

I don't make the rules :wink:

**Comment by hndgzkn** [comment]
> sklearn convention: estimator must be instantiable by default, so all parameters must have a default
> 
> I don't make the rules 😉

I cannot judge the convenience. However I have this question in mind: If `fit` cannot run with the default value of `labels_img`, could it actually be `y` parameter for `fit` which is omitted in `NiftiLabelsMasker`?

**Comment by bthirion** [comment]
No, to me `y` really means "the target", which is something else than the labels image.

**Comment by Remi-Gau** [comment]
@hndgzkn 
we should probably open a separate issue to discuss the transform behavior of nifti label or map maskers and whether we want to stick to sklearn rules (no change of the estiamtor by transform) or not
